File: libs/power-log-parser/src/lib/parsers/minion-on-board-attack-updated-parser.ts
import { GameTag, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { Action, Node, NodeType, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class MinionOnBoardAttackUpdatedParser implements ActionParser {
	readonly ParserName = 'MinionOnBoardAttackUpdatedParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === NodeType.TagChange &&
			(node.Object as TagChange).Name === (GameTag.ATK as number) &&
			this.GameState.CurrentEntities.get((node.Object as TagChange).Entity)!.GetTag(GameTag.ZONE) ===
				(Zone.PLAY as number)
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity)!;
		const initialAttack = entity.GetTag(GameTag.ATK);
		const newAttack = tagChange.Value;
		const cardId = entity.CardId;
		const controllerId = entity.GetEffectiveController();
		let sourceEntityId: number | null = null;
		let sourceCardId: string | null = null;
		if (node.Parent != null && node.Parent.Object instanceof Action) {
			const parentAction = node.Parent.Object as Action;
			sourceEntityId = parentAction.Entity;
			sourceCardId = this.GameState.CurrentEntities.get(parentAction.Entity)?.CardId ?? null;
		}
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'MINION_ON_BOARD_ATTACK_UPDATED',
				GameEventHelper.CreateProvider(
					'MINION_ON_BOARD_ATTACK_UPDATED',
					cardId,
					controllerId,
					entity.Id,
					this.StateFacade,
					{
						InitialAttack: initialAttack,
						NewAttack: newAttack,
						SourceEntityId: sourceEntityId,
						SourceCardId: sourceCardId,
					},
				),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/MinionOnBoardAttackUpdatedParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;

namespace HearthstoneReplays.Events.Parsers
{
    public class MinionOnBoardAttackUpdatedParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public MinionOnBoardAttackUpdatedParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.ATK
                && GameState.CurrentEntities[(node.Object as TagChange).Entity].GetTag(GameTag.ZONE) == (int)Zone.PLAY;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities[tagChange.Entity];
            var initialAttack = entity.GetTag(GameTag.ATK);
            var newAttack = tagChange.Value;
            var cardId = entity.CardId;
            var controllerId = entity.GetEffectiveController();
            //var gameState = GameEvent.BuildGameState(ParserState, StateFacade, GameState, tagChange, null);
            int? sourceEntityId = null;
            string sourceCardId = null;
            if (node.Parent != null && node.Parent.Object is Action)
            {
                var parentAction = node.Parent.Object as Action;
                sourceEntityId = parentAction.Entity;
                sourceCardId = GameState.CurrentEntities.GetValueOrDefault(parentAction.Entity)?.CardId;
            }
            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                "MINION_ON_BOARD_ATTACK_UPDATED",
                GameEvent.CreateProvider(
                    "MINION_ON_BOARD_ATTACK_UPDATED",
                    cardId,
                    controllerId,
                    entity.Id,
                    StateFacade,
                    //gameState,
                    new {
                        InitialAttack = initialAttack,
                        NewAttack = newAttack,
                        SourceEntityId = sourceEntityId,
                        SourceCardId = sourceCardId,
                    }
                ),
                true,
                node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
